KAFKA-16363: Storage tool crashes if dir is unavailable

mwesterby opened the pull request:
Prevents the storage tool from crashing when some of the directories are unavailable, but others are still available and have not yet been formatted.

Should all directories be unavailable however, and no existing directories have been formatted, then the tool will fail.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

mwesterby committed:
Prevent storage tool crashing when directory is unavailable

Prevents the storage tool from crashing when some of the
directories are unavailable, but others are still available and
have not yet been formatted.

chia7712 commented on core/src/main/scala/kafka/tools/StorageTool.scala:
```diff
@@ -440,8 +440,12 @@ object StorageTool extends Logging {
+        stream.println(s"I/O error trying to read log directory $errorLogDir. Ignoring...")
+      })
+      if(metaPropertiesEnsemble.emptyLogDirs().isEmpty) {
+        throw new TerseFailure("No available log directories to format.")
```
Do we need to check `logDirProps` before throwing this exception?

mwesterby replied:
Discussed below. I've now added the additional check to see if `logDirProps` is empty, and only throw this failure when this is also true. This is to prevent the format storage script from failing when the only remaining directories to be formatted are unavailable, but there is at least one directory already formatted, and the script is ran with `--ignore-formatted`.

In this scenario, it should succeed as there is at least one successfully formatted directory.

soarez reviewed (commented):
Thanks for the change.

It seems the three new tests have a lot in common:

* the setup of `MetaProperties`, `ByteArrayOutputStream` and `BootstrapMetadata`
* the invocation of command `formatCommand`

Is it worth extracting that into a utility method?

chia7712 reviewed (commented):
@mwesterby thanks for this patch.

one question: It seems `LogManager` (https://github.com/apache/kafka/blob/fccd7fec666d6570758e0b7891771099240ceee8/core/src/main/scala/kafka/log/LogManager.scala#L180) can't accept failed directories in startup, right? If so, is there a jira for that issue? I browse https://issues.apache.org/jira/browse/KAFKA-16061 and I failed to get a jira.

chia7712 commented:
> LogManager can accept failed directories in startup. That method is a bit confusing, but if you look a bit below the line you pointed to, the exception is caught, the fact the directory is failed is processed, and the directory is not added to liveLogDirs.

thanks for sharing this to me. happy to learn it :)

mwesterby committed:
Abstract common test setup to utility method

Abstracts the setup and invocation of the storage tool format command
into a utility method for new tests to simplify the testing logic and
reduce repetitive code.

mwesterby commented:
> It seems the three new tests have a lot in common:
>
> the setup of MetaProperties, ByteArrayOutputStream and BootstrapMetadata
the invocation of command formatCommand
Is it worth extracting that into a utility method?

@soarez I've refactored the tests to use a new utility method which prepares and invokes the format storage command, taking only the directories to format and the stream to use, which has cleaned up the common setup in the tests.

chia7712 commented on core/src/test/scala/unit/kafka/tools/StorageToolTest.scala:
```diff
@@ -192,6 +192,66 @@ Found problem:
+      assertTrue(stream.toString().contains("Formatting %s".format(availableDir1)))
+      assertFalse(stream.toString().contains("Formatting %s".format(unavailableDir1)))
+    } finally {
+      Utils.delete(availableDir1)
```
the file created by `TestUtils.tempFile()` will be delete on jvm exit (https://github.com/apache/kafka/blob/trunk/clients/src/test/java/org/apache/kafka/test/TestUtils.java#L198), so we don't need to call delete here

mwesterby replied:
Good spot! I'll remove these.

chia7712 commented on core/src/test/scala/unit/kafka/tools/StorageToolTest.scala:
```diff
@@ -192,6 +192,66 @@ Found problem:
+      try {
+        assertEquals(1, runFormatCommand(stream, Seq(unavailableDir1.toString, unavailableDir2.toString)))
+      } catch {
+        case e: TerseFailure => assertEquals("No available log directories to format.", e.getMessage)
```
it seems `runFormatCommand` is expected to be failed. We can use `assertThrows` to replace `try-catch`.

mwesterby replied:
Much cleaner, I'll refactor this assertion too.

chia7712 reviewed (commented):
@mwesterby thanks for updated PR. I leave some comments. Also, is my previous comment (https://github.com/apache/kafka/pull/15733#discussion_r1567431572) valid? the following test case is based on my comment, and it seems to me that should pass.

```scala
  @Test
  def testFormatSucceedsIfAtLeastOneFormattedDirectoryIsAvailable(): Unit = {
    val availableDir1 = TestUtils.tempDir()
    val stream = new ByteArrayOutputStream()
    assertEquals(0, runFormatCommand(stream, Seq(availableDir1.toString)))

    val stream2 = new ByteArrayOutputStream()
    val unavailableDir1 = TestUtils.tempFile()
    // availableDir1 is formatted and it is available. it should pass when `--ignore-formatted` is configured to true 
    assertEquals(0, runFormatCommand(stream2, Seq(availableDir1.toString, unavailableDir1.toString), ignoreFormatted = true))
  }
```
please feel free to correct me if I misunderstand anything.

mwesterby commented:
> Also, is my previous comment (https://github.com/apache/kafka/pull/15733#discussion_r1567431572) valid? the following test case is based on my comment, and it seems to me that should pass.

@chia7712 You are indeed correct! Apologies, I was still running through the code to figure it out but came to the same conclusion. We should not fail in this scenario, I'll update the PR now.

mwesterby committed:
Check no formatted directories exist before failing

Adds a check to determine if there are any existing formatted
log directories, and if so, not to fail the format command.
This is to prevent the command from failing when the only unformatted
directories that remain are unavailable, but there is already at
least one existing formatted directory, and it is ran with
--ignore-formatted.

Additionally, cleans up some of the tests and adds a new test for
this functionality.

mwesterby commented:
@chia7712 Thanks again for the review. I've implemented all of your suggested. Please let me know if you spot anything else I've missed.

chia7712 reviewed (approved):
LGTM

chia7712 commented:
@soarez it would be great to get your reviews before merging it :)

soarez commented:
Thanks @chia7712 I'll have another look at this later today

fvaleri reviewed (approved):
LGTM. Thanks.

soarez reviewed (approved):
LGTM thanks for the changes!
